// dirham/dist/index.js
// src/core/constants.ts
var DIRHAM_UNICODE = "\u20C3";
var DIRHAM_HTML_ENTITY = "&#x20C3;";
var DIRHAM_CSS_CONTENT = "\\20C3";
var DIRHAM_CURRENCY_CODE = "AED";
var DIRHAM_SYMBOL_TEXT = "\u062F.\u0625";
var DIRHAM_FONT_FAMILY = "Dirham";
var DIRHAM_CSS_CLASS = "dirham-symbol";
var DIRHAM_CODEPOINT = 8387;
var DIRHAM_WEIGHT_MAP = {
  thin: 100,
  extralight: 200,
  light: 300,
  regular: 400,
  medium: 500,
  semibold: 600,
  bold: 700,
  extrabold: 800,
  black: 900
};
var DIRHAM_STROKE_MAP = {
  thin: 0,
  extralight: 0,
  light: 0,
  regular: 0,
  medium: 8,
  semibold: 16,
  bold: 24,
  extrabold: 36,
  black: 48
};

// src/core/format.ts
var _fmtCache = /* @__PURE__ */ new Map();
function getFormatter(locale, decimals, notation = "standard") {
  const key = `${locale}:${decimals}:${notation}`;
  let fmt = _fmtCache.get(key);
  if (!fmt) {
    fmt = new Intl.NumberFormat(locale, {
      minimumFractionDigits: notation === "compact" ? 0 : decimals,
      maximumFractionDigits: decimals,
      ...notation === "compact" && {
        notation: "compact",
        compactDisplay: "short"
      }
    });
    _fmtCache.set(key, fmt);
  }
  return fmt;
}
function formatDirham(amount, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `formatDirham: amount must be a finite number, got ${amount}`
    );
  }
  const {
    locale = "en-AE",
    decimals = 2,
    useCode = false,
    separator = "\xA0",
    // non-breaking space
    notation = "standard"
  } = options;
  const symbol = useCode ? DIRHAM_CURRENCY_CODE : DIRHAM_UNICODE;
  let symbolFirst;
  if (options.symbolFirst !== void 0) {
    symbolFirst = options.symbolFirst;
  } else {
    symbolFirst = !locale.startsWith("ar");
  }
  const formatted = getFormatter(locale, decimals, notation).format(amount);
  return symbolFirst ? `${symbol}${separator}${formatted}` : `${formatted}${separator}${symbol}`;
}
function parseDirham(value, options = {}) {
  const { normalizeArabicNumerals = true } = options;
  let cleaned = value.replaceAll(DIRHAM_UNICODE, "").replaceAll(DIRHAM_SYMBOL_TEXT, "").replaceAll(DIRHAM_CURRENCY_CODE, "").replace(/[,\s\u00A0\u066C]/g, "").trim();
  if (normalizeArabicNumerals) {
    cleaned = cleaned.replace(
      /[\u0660-\u0669]/g,
      (d) => String(d.charCodeAt(0) - 1632)
    );
    cleaned = cleaned.replace(/\u066B/g, ".");
  }
  const result = Number.parseFloat(cleaned);
  if (Number.isNaN(result)) {
    throw new Error(`Cannot parse "${value}" as a Dirham amount`);
  }
  return result;
}

// src/core/clipboard.ts
var FORMAT_MAP = {
  unicode: DIRHAM_UNICODE,
  html: DIRHAM_HTML_ENTITY,
  css: DIRHAM_CSS_CONTENT,
  arabic: DIRHAM_SYMBOL_TEXT
};
async function copyDirhamSymbol(format = "unicode") {
  const text = FORMAT_MAP[format];
  if (typeof navigator === "undefined" || !navigator.clipboard) {
    throw new Error(
      "copyDirhamSymbol: Clipboard API is not available in this environment"
    );
  }
  await navigator.clipboard.writeText(text);
}

// src/core/copyAmount.ts
async function copyDirhamAmount(amount, options = {}) {
  if (typeof navigator === "undefined" || !navigator.clipboard) {
    throw new Error(
      "copyDirhamAmount: Clipboard API is not available in this environment"
    );
  }
  const text = formatDirham(amount, options);
  await navigator.clipboard.writeText(text);
}

// src/core/vat.ts
var UAE_VAT_RATE = 0.05;
function addVAT(amount, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `addVAT: amount must be a finite number, got ${amount}`
    );
  }
  const { rate = UAE_VAT_RATE, decimals = 2 } = options;
  const factor = 10 ** decimals;
  return Math.round(amount * (1 + rate) * factor) / factor;
}
function removeVAT(amount, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `removeVAT: amount must be a finite number, got ${amount}`
    );
  }
  const { rate = UAE_VAT_RATE, decimals = 2 } = options;
  const factor = 10 ** decimals;
  return Math.round(amount / (1 + rate) * factor) / factor;
}
function getVAT(amount, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `getVAT: amount must be a finite number, got ${amount}`
    );
  }
  const { rate = UAE_VAT_RATE, decimals = 2 } = options;
  const factor = 10 ** decimals;
  return Math.round(amount * rate * factor) / factor;
}

// src/core/conversion.ts
var _cache = null;
var CACHE_TTL = 60 * 60 * 1e3;
async function fetchExchangeRates() {
  if (_cache && Date.now() - _cache.timestamp < CACHE_TTL) {
    return _cache.rates;
  }
  const res = await fetch("https://open.er-api.com/v6/latest/AED");
  if (!res.ok) {
    throw new Error(
      `Failed to fetch exchange rates: ${res.status} ${res.statusText}`
    );
  }
  const data = await res.json();
  if (!data.rates) {
    throw new Error("Invalid exchange rate response: missing rates");
  }
  _cache = { timestamp: Date.now(), rates: data.rates };
  return data.rates;
}
function clearRateCache() {
  _cache = null;
}
async function convertFromAED(amount, currency, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `convertFromAED: amount must be a finite number, got ${amount}`
    );
  }
  const { decimals = 2 } = options;
  const code = currency.toUpperCase();
  let rate;
  if (options.rate !== void 0) {
    rate = options.rate;
  } else {
    const rates = await fetchExchangeRates();
    const r = rates[code];
    if (r === void 0) {
      throw new Error(`Unknown currency code: ${code}`);
    }
    rate = r;
  }
  const factor = 10 ** decimals;
  return Math.round(amount * rate * factor) / factor;
}
async function convertToAED(amount, currency, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `convertToAED: amount must be a finite number, got ${amount}`
    );
  }
  const { decimals = 2 } = options;
  const code = currency.toUpperCase();
  let rate;
  if (options.rate !== void 0) {
    rate = options.rate;
  } else {
    const rates = await fetchExchangeRates();
    const r = rates[code];
    if (r === void 0) {
      throw new Error(`Unknown currency code: ${code}`);
    }
    rate = r;
  }
  if (rate === 0) {
    throw new Error(`Exchange rate for ${code} is zero`);
  }
  const factor = 10 ** decimals;
  return Math.round(amount / rate * factor) / factor;
}

export { DIRHAM_CODEPOINT, DIRHAM_CSS_CLASS, DIRHAM_CSS_CONTENT, DIRHAM_CURRENCY_CODE, DIRHAM_FONT_FAMILY, DIRHAM_HTML_ENTITY, DIRHAM_STROKE_MAP, DIRHAM_SYMBOL_TEXT, DIRHAM_UNICODE, DIRHAM_WEIGHT_MAP, UAE_VAT_RATE, addVAT, clearRateCache, convertFromAED, convertToAED, copyDirhamAmount, copyDirhamSymbol, fetchExchangeRates, formatDirham, getVAT, parseDirham, removeVAT };
