// dirham/dist/react/index.cjs
'use strict';

var react = require('react');
var jsxRuntime = require('react/jsx-runtime');

// src/react/DirhamSymbol.tsx

// src/core/constants.ts
var DIRHAM_UNICODE = "\u20C3";
var DIRHAM_CURRENCY_CODE = "AED";
var DIRHAM_STROKE_MAP = {
  thin: 0,
  extralight: 0,
  light: 0,
  regular: 0,
  medium: 8,
  semibold: 16,
  bold: 24,
  extrabold: 36,
  black: 48
};
var DirhamSymbolBase = react.forwardRef(
  ({
    size = 24,
    color = "currentColor",
    "aria-label": ariaLabel = "UAE Dirham",
    weight = "regular",
    style,
    ...props
  }, ref) => {
    const strokeWidth = DIRHAM_STROKE_MAP[weight];
    return /* @__PURE__ */ jsxRuntime.jsx(
      "svg",
      {
        ref,
        xmlns: "http://www.w3.org/2000/svg",
        viewBox: "0 0 1000 870",
        width: size,
        height: size,
        fill: color,
        role: "img",
        "aria-label": ariaLabel,
        style: {
          display: "inline-block",
          verticalAlign: "-0.125em",
          ...style
        },
        ...props,
        children: /* @__PURE__ */ jsxRuntime.jsx(
          "path",
          {
            d: "m88.3 1c0.4 0.6 2.6 3.3 4.7 5.9 15.3 18.2 26.8 47.8 33 85.1 4.1 24.5 4.3 32.2 4.3 125.6v87h-41.8c-38.2 0-42.6-0.2-50.1-1.7-11.8-2.5-24-9.2-32.2-17.8-6.5-6.9-6.3-7.3-5.9 13.6 0.5 17.3 0.7 19.2 3.2 28.6 4 14.9 9.5 26 17.8 35.9 11.3 13.6 22.8 21.2 39.2 26.3 3.5 1 10.9 1.4 37.1 1.6l32.7 0.5v43.3 43.4l-46.1-0.3-46.3-0.3-8-3.2c-9.5-3.8-13.8-6.6-23.1-14.9l-6.8-6.1 0.4 19.1c0.5 17.7 0.6 19.7 3.1 28.7 8.7 31.8 29.7 54.5 57.4 61.9 6.9 1.9 9.6 2 38.5 2.4l30.9 0.4v89.6c0 54.1-0.3 94-0.8 100.8-0.5 6.2-2.1 17.8-3.5 25.9-6.5 37.3-18.2 65.4-35 83.6l-3.4 3.7h169.1c101.1 0 176.7-0.4 187.8-0.9 19.5-1 63-5.3 72.8-7.4 3.1-0.6 8.9-1.5 12.7-2.1 8.1-1.2 21.5-4 40.8-8.9 27.2-6.8 52-15.3 76.3-26.1 7.6-3.4 29.4-14.5 35.2-18 3.1-1.8 6.8-4 8.2-4.7 3.9-2.1 10.4-6.3 19.9-13.1 4.7-3.4 9.4-6.7 10.4-7.4 4.2-2.8 18.7-14.9 25.3-21 25.1-23.1 46.1-48.8 62.4-76.3 2.3-4 5.3-9 6.6-11.1 3.3-5.6 16.9-33.6 18.2-37.8 0.6-1.9 1.4-3.9 1.8-4.3 2.6-3.4 17.6-50.6 19.4-60.9 0.6-3.3 0.9-3.8 3.4-4.3 1.6-0.3 24.9-0.3 51.8-0.1 53.8 0.4 53.8 0.4 65.7 5.9 6.7 3.1 8.7 4.5 16.1 11.2 9.7 8.7 8.8 10.1 8.2-11.7-0.4-12.8-0.9-20.7-1.8-23.9-3.4-12.3-4.2-14.9-7.2-21.1-9.8-21.4-26.2-36.7-47.2-44l-8.2-3-33.4-0.4-33.3-0.5 0.4-11.7c0.4-15.4 0.4-45.9-0.1-61.6l-0.4-12.6 44.6-0.2c38.2-0.2 45.3 0 49.5 1.1 12.6 3.5 21.1 8.3 31.5 17.8l5.8 5.4v-14.8c0-17.6-0.9-25.4-4.5-37-7.1-23.5-21.1-41-41.1-51.8-13-7-13.8-7.2-58.5-7.5-26.2-0.2-39.9-0.6-40.6-1.2-0.6-0.6-1.1-1.6-1.1-2.4 0-0.8-1.5-7.1-3.5-13.9-23.4-82.7-67.1-148.4-131-197.1-8.7-6.7-30-20.8-38.6-25.6-3.3-1.9-6.9-3.9-7.8-4.5-4.2-2.3-28.3-14.1-34.3-16.6-3.6-1.6-8.3-3.6-10.4-4.4-35.3-15.3-94.5-29.8-139.7-34.3-7.4-0.7-17.2-1.8-21.7-2.2-20.4-2.3-48.7-2.6-209.4-2.6-135.8 0-169.9 0.3-169.4 1zm330.7 43.3c33.8 2 54.6 4.6 78.9 10.5 74.2 17.6 126.4 54.8 164.3 117 3.5 5.8 18.3 36 20.5 42.1 10.5 28.3 15.6 45.1 20.1 67.3 1.1 5.4 2.6 12.6 3.3 16 0.7 3.3 1 6.4 0.7 6.7-0.5 0.4-100.9 0.6-223.3 0.5l-222.5-0.2-0.3-128.5c-0.1-70.6 0-129.3 0.3-130.4l0.4-1.9h71.1c39 0 78 0.4 86.5 0.9zm297.5 350.3c0.7 4.3 0.7 77.3 0 80.9l-0.6 2.7-227.5-0.2-227.4-0.3-0.2-42.4c-0.2-23.3 0-42.7 0.2-43.1 0.3-0.5 97.2-0.8 227.7-0.8h227.2zm-10.2 171.7c0.5 1.5-1.9 13.8-6.8 33.8-5.6 22.5-13.2 45.2-20.9 62-3.8 8.6-13.3 27.2-15.6 30.7-1.1 1.6-4.3 6.7-7.1 11.2-18 28.2-43.7 53.9-73 72.9-10.7 6.8-32.7 18.4-38.6 20.2-1.2 0.3-2.5 0.9-3 1.3-0.7 0.6-9.8 4-20.4 7.8-19.5 6.9-56.6 14.4-86.4 17.5-19.3 1.9-22.4 2-96.7 2h-76.9v-129.7-129.8l220.9-0.4c121.5-0.2 221.6-0.5 222.4-0.7 0.9-0.1 1.8 0.5 2.1 1.2z",
            ...strokeWidth > 0 && {
              stroke: color,
              strokeWidth,
              strokeLinejoin: "round",
              paintOrder: "stroke"
            }
          }
        )
      }
    );
  }
);
DirhamSymbolBase.displayName = "DirhamSymbol";
var DirhamSymbol = react.memo(DirhamSymbolBase);
DirhamSymbol.displayName = "DirhamSymbol";
var DirhamIconBase = react.forwardRef(
  ({
    size,
    color = "currentColor",
    "aria-label": ariaLabel = "UAE Dirham",
    as: Tag = "i",
    className = "",
    style,
    ...props
  }, ref) => {
    const classes = `dirham-symbol${className ? ` ${className}` : ""}`;
    const iconStyle = {
      ...size !== void 0 && {
        fontSize: typeof size === "number" ? `${size}px` : size
      },
      color,
      ...style
    };
    return /* @__PURE__ */ jsxRuntime.jsx(
      Tag,
      {
        ref,
        className: classes,
        style: iconStyle,
        role: "img",
        "aria-label": ariaLabel,
        ...props
      }
    );
  }
);
DirhamIconBase.displayName = "DirhamIcon";
var DirhamIcon = react.memo(DirhamIconBase);
DirhamIcon.displayName = "DirhamIcon";

// src/core/format.ts
var _fmtCache = /* @__PURE__ */ new Map();
function getFormatter(locale, decimals, notation = "standard") {
  const key = `${locale}:${decimals}:${notation}`;
  let fmt = _fmtCache.get(key);
  if (!fmt) {
    fmt = new Intl.NumberFormat(locale, {
      minimumFractionDigits: notation === "compact" ? 0 : decimals,
      maximumFractionDigits: decimals,
      ...notation === "compact" && {
        notation: "compact",
        compactDisplay: "short"
      }
    });
    _fmtCache.set(key, fmt);
  }
  return fmt;
}
function formatDirham(amount, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `formatDirham: amount must be a finite number, got ${amount}`
    );
  }
  const {
    locale = "en-AE",
    decimals = 2,
    useCode = false,
    separator = "\xA0",
    // non-breaking space
    notation = "standard"
  } = options;
  const symbol = useCode ? DIRHAM_CURRENCY_CODE : DIRHAM_UNICODE;
  let symbolFirst;
  if (options.symbolFirst !== void 0) {
    symbolFirst = options.symbolFirst;
  } else {
    symbolFirst = !locale.startsWith("ar");
  }
  const formatted = getFormatter(locale, decimals, notation).format(amount);
  return symbolFirst ? `${symbol}${separator}${formatted}` : `${formatted}${separator}${symbol}`;
}
var DirhamPriceBase = react.forwardRef(
  ({
    amount,
    locale = "en-AE",
    decimals = 2,
    useCode = false,
    notation = "standard",
    symbolSize = "1em",
    weight = "regular",
    "aria-label": ariaLabel,
    style,
    className,
    ...props
  }, ref) => {
    const symbolFirst = !locale.startsWith("ar");
    const formatted = react.useMemo(() => {
      if (!Number.isFinite(amount)) return "\u2014";
      return formatDirham(amount, {
        locale,
        decimals,
        useCode: true,
        // Always use code so we can strip it and render SVG instead
        notation
      }).replace("AED", "").trim();
    }, [amount, locale, decimals, notation]);
    const symbol = useCode ? /* @__PURE__ */ jsxRuntime.jsx("span", { children: "AED" }) : /* @__PURE__ */ jsxRuntime.jsx(
      DirhamSymbol,
      {
        size: symbolSize,
        weight,
        "aria-label": ariaLabel ?? "UAE Dirham"
      }
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      "span",
      {
        ref,
        className,
        dir: locale.startsWith("ar") ? "rtl" : void 0,
        style: { whiteSpace: "nowrap", ...style },
        ...props,
        children: symbolFirst ? /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          symbol,
          "\xA0",
          formatted
        ] }) : /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          formatted,
          "\xA0",
          symbol
        ] })
      }
    );
  }
);
DirhamPriceBase.displayName = "DirhamPrice";
var DirhamPrice = react.memo(DirhamPriceBase);
DirhamPrice.displayName = "DirhamPrice";
var _fmtCache2 = /* @__PURE__ */ new Map();
function getInputFormatter(locale, decimals) {
  const key = `input:${locale}:${decimals}`;
  let fmt = _fmtCache2.get(key);
  if (!fmt) {
    fmt = new Intl.NumberFormat(locale, {
      minimumFractionDigits: 0,
      maximumFractionDigits: decimals
    });
    _fmtCache2.set(key, fmt);
  }
  return fmt;
}
function formatInputValue(value, locale, decimals) {
  if (value === void 0 || !Number.isFinite(value)) return "";
  return getInputFormatter(locale, decimals).format(value);
}
function stripFormatting(raw) {
  let cleaned = raw.replace(
    /[\u0660-\u0669]/g,
    (d) => String(d.charCodeAt(0) - 1632)
  );
  cleaned = cleaned.replace(/\u066B/g, ".");
  cleaned = cleaned.replace(/[,\s\u00A0\u066C]/g, "");
  cleaned = cleaned.replace(/[\u20C3]/g, "");
  cleaned = cleaned.replace(/AED/gi, "");
  cleaned = cleaned.replace(/د\.إ/g, "");
  return cleaned.trim();
}
var DirhamInputBase = react.forwardRef(
  ({
    value: controlledValue,
    defaultValue,
    onChange,
    locale = "en-AE",
    decimals = 2,
    min,
    max,
    showSymbol = true,
    useCode = false,
    symbolSize = "1em",
    weight = "regular",
    "aria-label": ariaLabel = "Amount in AED",
    className,
    style,
    inputClassName,
    inputStyle,
    onBlur,
    onFocus,
    selectOnFocus = true,
    placeholder,
    disabled,
    readOnly,
    ...inputProps
  }, ref) => {
    const isControlled = controlledValue !== void 0;
    const [internalValue, setInternalValue] = react.useState(
      defaultValue
    );
    const numericValue = isControlled ? controlledValue : internalValue;
    const [rawText, setRawText] = react.useState("");
    const [isFocused, setIsFocused] = react.useState(false);
    const inputRef = react.useRef(null);
    const mergedRef = react.useCallback(
      (node) => {
        inputRef.current = node;
        if (typeof ref === "function") ref(node);
        else if (ref)
          ref.current = node;
      },
      [ref]
    );
    const isRTL = locale.startsWith("ar");
    const clamp = react.useCallback(
      (v) => {
        let result = v;
        if (min !== void 0 && result < min) result = min;
        if (max !== void 0 && result > max) result = max;
        return result;
      },
      [min, max]
    );
    const displayText = isFocused ? rawText : formatInputValue(numericValue, locale, decimals);
    const handleFocus = react.useCallback(
      (e) => {
        setIsFocused(true);
        const plain = numericValue !== void 0 && Number.isFinite(numericValue) ? numericValue.toString() : "";
        setRawText(plain);
        if (selectOnFocus) {
          setTimeout(() => inputRef.current?.select(), 0);
        }
        onFocus?.(e);
      },
      [numericValue, selectOnFocus, onFocus]
    );
    const handleBlur = react.useCallback(
      (e) => {
        setIsFocused(false);
        const stripped = stripFormatting(rawText);
        if (stripped === "" || stripped === "." || stripped === "-") {
          if (!isControlled) setInternalValue(void 0);
          onChange?.(void 0);
        } else {
          let parsed = Number.parseFloat(stripped);
          if (Number.isNaN(parsed)) {
            if (!isControlled) setInternalValue(void 0);
            onChange?.(void 0);
          } else {
            const factor = 10 ** decimals;
            parsed = Math.round(parsed * factor) / factor;
            parsed = clamp(parsed);
            if (!isControlled) setInternalValue(parsed);
            onChange?.(parsed);
          }
        }
        onBlur?.(e);
      },
      [rawText, decimals, clamp, isControlled, onChange, onBlur]
    );
    const handleChange = react.useCallback(
      (e) => {
        const text = e.target.value;
        const cleaned = stripFormatting(text);
        if (cleaned !== "" && !/^-?\d*\.?\d*$/.test(cleaned)) {
          return;
        }
        const dotIndex = cleaned.indexOf(".");
        if (dotIndex !== -1) {
          const decimalPart = cleaned.slice(dotIndex + 1);
          if (decimalPart.length > decimals) {
            return;
          }
        }
        setRawText(cleaned);
        if (cleaned === "" || cleaned === "." || cleaned === "-") {
          if (!isControlled) setInternalValue(void 0);
          onChange?.(void 0);
        } else {
          const parsed = Number.parseFloat(cleaned);
          if (!Number.isNaN(parsed)) {
            if (!isControlled) setInternalValue(parsed);
            onChange?.(parsed);
          }
        }
      },
      [decimals, isControlled, onChange]
    );
    const handlePaste = react.useCallback(
      (e) => {
        e.preventDefault();
        const pasted = e.clipboardData.getData("text/plain");
        const stripped = stripFormatting(pasted);
        if (stripped === "") return;
        const dotIndex = stripped.indexOf(".");
        let final = stripped;
        if (dotIndex !== -1) {
          final = `${stripped.slice(0, dotIndex)}.${stripped.slice(dotIndex + 1, dotIndex + 1 + decimals)}`;
        }
        if (!/^-?\d*\.?\d*$/.test(final)) return;
        setRawText(final);
        const parsed = Number.parseFloat(final);
        if (!Number.isNaN(parsed)) {
          const clamped = clamp(
            Math.round(parsed * 10 ** decimals) / 10 ** decimals
          );
          if (!isControlled) setInternalValue(clamped);
          onChange?.(clamped);
        }
      },
      [decimals, clamp, isControlled, onChange]
    );
    const symbolElement = showSymbol ? useCode ? /* @__PURE__ */ jsxRuntime.jsx(
      "span",
      {
        style: {
          userSelect: "none",
          flexShrink: 0,
          ...isRTL ? { marginLeft: "0.5em" } : { marginRight: "0.5em" }
        },
        "aria-hidden": "true",
        children: "AED"
      }
    ) : /* @__PURE__ */ jsxRuntime.jsx(
      DirhamSymbol,
      {
        size: symbolSize,
        weight,
        "aria-hidden": "true",
        style: {
          flexShrink: 0,
          ...isRTL ? { marginLeft: "0.5em" } : { marginRight: "0.5em" }
        }
      }
    ) : null;
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "span",
      {
        className,
        style: {
          display: "inline-flex",
          alignItems: "center",
          direction: isRTL ? "rtl" : "ltr",
          ...style
        },
        children: [
          !isRTL && symbolElement,
          /* @__PURE__ */ jsxRuntime.jsx(
            "input",
            {
              ...inputProps,
              ref: mergedRef,
              type: "text",
              inputMode: "decimal",
              dir: isRTL ? "rtl" : "ltr",
              "aria-label": ariaLabel,
              className: inputClassName,
              style: {
                minWidth: 0,
                flex: 1,
                ...inputStyle
              },
              value: displayText,
              placeholder: placeholder ?? formatInputValue(0, locale, decimals),
              disabled,
              readOnly,
              onChange: handleChange,
              onFocus: handleFocus,
              onBlur: handleBlur,
              onPaste: handlePaste
            }
          ),
          isRTL && symbolElement
        ]
      }
    );
  }
);
DirhamInputBase.displayName = "DirhamInput";
var DirhamInput = react.memo(DirhamInputBase);
DirhamInput.displayName = "DirhamInput";
var EASING_FNS = {
  linear: (t) => t,
  easeIn: (t) => t * t,
  easeOut: (t) => t * (2 - t),
  easeInOut: (t) => t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t
};
var AnimatedDirhamPriceBase = react.forwardRef(
  ({
    amount,
    duration = 600,
    easing = "easeOut",
    locale = "en-AE",
    decimals = 2,
    useCode = false,
    notation = "standard",
    symbolSize = "1em",
    weight = "regular",
    "aria-label": ariaLabel,
    style,
    className,
    ...props
  }, ref) => {
    const [displayAmount, setDisplayAmount] = react.useState(amount);
    const prevAmountRef = react.useRef(amount);
    const frameRef = react.useRef(0);
    react.useEffect(() => {
      const from = prevAmountRef.current;
      const to = amount;
      prevAmountRef.current = to;
      if (from === to) return;
      const easeFn = EASING_FNS[easing] ?? EASING_FNS.easeOut;
      const start = performance.now();
      const animate = (now) => {
        const elapsed = now - start;
        const progress = Math.min(elapsed / duration, 1);
        const eased = easeFn(progress);
        const current = from + (to - from) * eased;
        setDisplayAmount(current);
        if (progress < 1) {
          frameRef.current = requestAnimationFrame(animate);
        } else {
          setDisplayAmount(to);
        }
      };
      frameRef.current = requestAnimationFrame(animate);
      return () => {
        if (frameRef.current) cancelAnimationFrame(frameRef.current);
      };
    }, [amount, duration, easing]);
    const symbolFirst = !locale.startsWith("ar");
    const isRTL = locale.startsWith("ar");
    const formatted = Number.isFinite(displayAmount) ? formatDirham(displayAmount, {
      locale,
      decimals,
      useCode: true,
      notation
    }).replace("AED", "").trim() : "\u2014";
    const symbol = useCode ? /* @__PURE__ */ jsxRuntime.jsx("span", { children: "AED" }) : /* @__PURE__ */ jsxRuntime.jsx(
      DirhamSymbol,
      {
        size: symbolSize,
        weight,
        "aria-label": ariaLabel ?? "UAE Dirham"
      }
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      "span",
      {
        ref,
        className,
        dir: isRTL ? "rtl" : void 0,
        style: { whiteSpace: "nowrap", ...style },
        "aria-live": "polite",
        "aria-atomic": "true",
        ...props,
        children: symbolFirst ? /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          symbol,
          "\xA0",
          formatted
        ] }) : /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          formatted,
          "\xA0",
          symbol
        ] })
      }
    );
  }
);
AnimatedDirhamPriceBase.displayName = "AnimatedDirhamPrice";
var AnimatedDirhamPrice = react.memo(AnimatedDirhamPriceBase);
AnimatedDirhamPrice.displayName = "AnimatedDirhamPrice";

// src/core/conversion.ts
var _cache = null;
var CACHE_TTL = 60 * 60 * 1e3;
async function fetchExchangeRates() {
  if (_cache && Date.now() - _cache.timestamp < CACHE_TTL) {
    return _cache.rates;
  }
  const res = await fetch("https://open.er-api.com/v6/latest/AED");
  if (!res.ok) {
    throw new Error(
      `Failed to fetch exchange rates: ${res.status} ${res.statusText}`
    );
  }
  const data = await res.json();
  if (!data.rates) {
    throw new Error("Invalid exchange rate response: missing rates");
  }
  _cache = { timestamp: Date.now(), rates: data.rates };
  return data.rates;
}
function clearRateCache() {
  _cache = null;
}

// src/react/useDirhamRate.ts
function useDirhamRate(currency) {
  const [rate, setRate] = react.useState(void 0);
  const [loading, setLoading] = react.useState(true);
  const [error, setError] = react.useState(void 0);
  const mountedRef = react.useRef(true);
  const code = currency.toUpperCase();
  const doFetch = react.useCallback(() => {
    setLoading(true);
    setError(void 0);
    fetchExchangeRates().then((rates) => {
      if (!mountedRef.current) return;
      const r = rates[code];
      if (r === void 0) {
        setError(`Unknown currency code: ${code}`);
        setRate(void 0);
      } else {
        setRate(r);
      }
    }).catch((err) => {
      if (!mountedRef.current) return;
      setError(err instanceof Error ? err.message : String(err));
      setRate(void 0);
    }).finally(() => {
      if (mountedRef.current) setLoading(false);
    });
  }, [code]);
  react.useEffect(() => {
    mountedRef.current = true;
    doFetch();
    return () => {
      mountedRef.current = false;
    };
  }, [doFetch]);
  const refetch = react.useCallback(() => {
    clearRateCache();
    doFetch();
  }, [doFetch]);
  const convert = react.useCallback(
    (amount) => {
      if (rate === void 0) return void 0;
      const factor = 100;
      return Math.round(amount * rate * factor) / factor;
    },
    [rate]
  );
  const convertBack = react.useCallback(
    (amount) => {
      if (rate === void 0 || rate === 0) return void 0;
      const factor = 100;
      return Math.round(amount / rate * factor) / factor;
    },
    [rate]
  );
  return { rate, loading, error, refetch, convert, convertBack };
}

exports.AnimatedDirhamPrice = AnimatedDirhamPrice;
exports.DirhamIcon = DirhamIcon;
exports.DirhamInput = DirhamInput;
exports.DirhamPrice = DirhamPrice;
exports.DirhamSymbol = DirhamSymbol;
exports.useDirhamRate = useDirhamRate;
