// dirham/dist/tailwind/index.cjs
'use strict';

var plugin = require('tailwindcss/plugin');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var plugin__default = /*#__PURE__*/_interopDefault(plugin);

// src/tailwind/index.ts
var dirhamPlugin = plugin__default.default(({ addUtilities, addComponents }) => {
  addUtilities({
    ".dirham": {
      fontFamily: "Dirham, sans-serif",
      fontStyle: "normal",
      display: "inline-block",
      verticalAlign: "middle"
    }
  });
  addUtilities({
    ".dirham-thin": { fontWeight: "100" },
    ".dirham-extralight": { fontWeight: "200" },
    ".dirham-light": { fontWeight: "300" },
    ".dirham-regular": { fontWeight: "400" },
    ".dirham-medium": { fontWeight: "500" },
    ".dirham-semibold": { fontWeight: "600" },
    ".dirham-bold": { fontWeight: "700" },
    ".dirham-extrabold": { fontWeight: "800" },
    ".dirham-black": { fontWeight: "900" }
  });
  addUtilities({
    ".dirham-xs": { fontSize: "0.75rem" },
    ".dirham-sm": { fontSize: "0.875rem" },
    ".dirham-base": { fontSize: "1rem" },
    ".dirham-lg": { fontSize: "1.125rem" },
    ".dirham-xl": { fontSize: "1.25rem" },
    ".dirham-2xl": { fontSize: "1.5rem" },
    ".dirham-3xl": { fontSize: "1.875rem" },
    ".dirham-4xl": { fontSize: "2.25rem" }
  });
  addUtilities({
    ".dirham-before::before": {
      content: '"\\20C3\\00A0"',
      fontFamily: "Dirham, sans-serif",
      fontStyle: "normal"
    },
    ".dirham-after::after": {
      content: '"\\00A0\\20C3"',
      fontFamily: "Dirham, sans-serif",
      fontStyle: "normal"
    }
  });
  addComponents({
    ".dirham-price": {
      whiteSpace: "nowrap",
      fontVariantNumeric: "tabular-nums"
    }
  });
});
var tailwind_default = dirhamPlugin;

module.exports = tailwind_default;
