// dirham/dist/web-component/index.cjs
'use strict';

var __typeError = (msg) => {
  throw TypeError(msg);
};
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);

// src/core/constants.ts
var DIRHAM_UNICODE = "\u20C3";
var DIRHAM_CURRENCY_CODE = "AED";
var DIRHAM_STROKE_MAP = {
  thin: 0,
  extralight: 0,
  light: 0,
  regular: 0,
  medium: 8,
  semibold: 16,
  bold: 24,
  extrabold: 36,
  black: 48
};

// src/web-component/DirhamSymbolElement.ts
var SVG_PATH = "m88.3 1c0.4 0.6 2.6 3.3 4.7 5.9 15.3 18.2 26.8 47.8 33 85.1 4.1 24.5 4.3 32.2 4.3 125.6v87h-41.8c-38.2 0-42.6-0.2-50.1-1.7-11.8-2.5-24-9.2-32.2-17.8-6.5-6.9-6.3-7.3-5.9 13.6 0.5 17.3 0.7 19.2 3.2 28.6 4 14.9 9.5 26 17.8 35.9 11.3 13.6 22.8 21.2 39.2 26.3 3.5 1 10.9 1.4 37.1 1.6l32.7 0.5v43.3 43.4l-46.1-0.3-46.3-0.3-8-3.2c-9.5-3.8-13.8-6.6-23.1-14.9l-6.8-6.1 0.4 19.1c0.5 17.7 0.6 19.7 3.1 28.7 8.7 31.8 29.7 54.5 57.4 61.9 6.9 1.9 9.6 2 38.5 2.4l30.9 0.4v89.6c0 54.1-0.3 94-0.8 100.8-0.5 6.2-2.1 17.8-3.5 25.9-6.5 37.3-18.2 65.4-35 83.6l-3.4 3.7h169.1c101.1 0 176.7-0.4 187.8-0.9 19.5-1 63-5.3 72.8-7.4 3.1-0.6 8.9-1.5 12.7-2.1 8.1-1.2 21.5-4 40.8-8.9 27.2-6.8 52-15.3 76.3-26.1 7.6-3.4 29.4-14.5 35.2-18 3.1-1.8 6.8-4 8.2-4.7 3.9-2.1 10.4-6.3 19.9-13.1 4.7-3.4 9.4-6.7 10.4-7.4 4.2-2.8 18.7-14.9 25.3-21 25.1-23.1 46.1-48.8 62.4-76.3 2.3-4 5.3-9 6.6-11.1 3.3-5.6 16.9-33.6 18.2-37.8 0.6-1.9 1.4-3.9 1.8-4.3 2.6-3.4 17.6-50.6 19.4-60.9 0.6-3.3 0.9-3.8 3.4-4.3 1.6-0.3 24.9-0.3 51.8-0.1 53.8 0.4 53.8 0.4 65.7 5.9 6.7 3.1 8.7 4.5 16.1 11.2 9.7 8.7 8.8 10.1 8.2-11.7-0.4-12.8-0.9-20.7-1.8-23.9-3.4-12.3-4.2-14.9-7.2-21.1-9.8-21.4-26.2-36.7-47.2-44l-8.2-3-33.4-0.4-33.3-0.5 0.4-11.7c0.4-15.4 0.4-45.9-0.1-61.6l-0.4-12.6 44.6-0.2c38.2-0.2 45.3 0 49.5 1.1 12.6 3.5 21.1 8.3 31.5 17.8l5.8 5.4v-14.8c0-17.6-0.9-25.4-4.5-37-7.1-23.5-21.1-41-41.1-51.8-13-7-13.8-7.2-58.5-7.5-26.2-0.2-39.9-0.6-40.6-1.2-0.6-0.6-1.1-1.6-1.1-2.4 0-0.8-1.5-7.1-3.5-13.9-23.4-82.7-67.1-148.4-131-197.1-8.7-6.7-30-20.8-38.6-25.6-3.3-1.9-6.9-3.9-7.8-4.5-4.2-2.3-28.3-14.1-34.3-16.6-3.6-1.6-8.3-3.6-10.4-4.4-35.3-15.3-94.5-29.8-139.7-34.3-7.4-0.7-17.2-1.8-21.7-2.2-20.4-2.3-48.7-2.6-209.4-2.6-135.8 0-169.9 0.3-169.4 1zm330.7 43.3c33.8 2 54.6 4.6 78.9 10.5 74.2 17.6 126.4 54.8 164.3 117 3.5 5.8 18.3 36 20.5 42.1 10.5 28.3 15.6 45.1 20.1 67.3 1.1 5.4 2.6 12.6 3.3 16 0.7 3.3 1 6.4 0.7 6.7-0.5 0.4-100.9 0.6-223.3 0.5l-222.5-0.2-0.3-128.5c-0.1-70.6 0-129.3 0.3-130.4l0.4-1.9h71.1c39 0 78 0.4 86.5 0.9zm297.5 350.3c0.7 4.3 0.7 77.3 0 80.9l-0.6 2.7-227.5-0.2-227.4-0.3-0.2-42.4c-0.2-23.3 0-42.7 0.2-43.1 0.3-0.5 97.2-0.8 227.7-0.8h227.2zm-10.2 171.7c0.5 1.5-1.9 13.8-6.8 33.8-5.6 22.5-13.2 45.2-20.9 62-3.8 8.6-13.3 27.2-15.6 30.7-1.1 1.6-4.3 6.7-7.1 11.2-18 28.2-43.7 53.9-73 72.9-10.7 6.8-32.7 18.4-38.6 20.2-1.2 0.3-2.5 0.9-3 1.3-0.7 0.6-9.8 4-20.4 7.8-19.5 6.9-56.6 14.4-86.4 17.5-19.3 1.9-22.4 2-96.7 2h-76.9v-129.7-129.8l220.9-0.4c121.5-0.2 221.6-0.5 222.4-0.7 0.9-0.1 1.8 0.5 2.1 1.2z";
var VALID_WEIGHTS = /* @__PURE__ */ new Set([
  "thin",
  "extralight",
  "light",
  "regular",
  "medium",
  "semibold",
  "bold",
  "extrabold",
  "black"
]);
var _shadow, _DirhamSymbolElement_instances, render_fn;
var DirhamSymbolElement = class extends HTMLElement {
  constructor() {
    super();
    __privateAdd(this, _DirhamSymbolElement_instances);
    __privateAdd(this, _shadow);
    __privateSet(this, _shadow, this.attachShadow({ mode: "open" }));
    __privateMethod(this, _DirhamSymbolElement_instances, render_fn).call(this);
  }
  static get observedAttributes() {
    return ["size", "color", "weight", "aria-label"];
  }
  attributeChangedCallback() {
    __privateMethod(this, _DirhamSymbolElement_instances, render_fn).call(this);
  }
};
_shadow = new WeakMap();
_DirhamSymbolElement_instances = new WeakSet();
render_fn = function() {
  const size = this.getAttribute("size") || "24";
  const color = this.getAttribute("color") || "currentColor";
  const weightAttr = this.getAttribute("weight") || "regular";
  const weight = VALID_WEIGHTS.has(weightAttr) ? weightAttr : "regular";
  const ariaLabel = this.getAttribute("aria-label") || "UAE Dirham";
  const strokeWidth = DIRHAM_STROKE_MAP[weight];
  const strokeAttrs = strokeWidth > 0 ? `stroke="${color}" stroke-width="${strokeWidth}" stroke-linejoin="round" paint-order="stroke"` : "";
  __privateGet(this, _shadow).innerHTML = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1000 870" width="${size}" height="${size}" fill="${color}" role="img" aria-label="${ariaLabel}" style="display:inline-block;vertical-align:middle"><path d="${SVG_PATH}" ${strokeAttrs}/></svg>`;
};
if (typeof customElements !== "undefined" && !customElements.get("dirham-symbol")) {
  customElements.define("dirham-symbol", DirhamSymbolElement);
}

// src/core/format.ts
var _fmtCache = /* @__PURE__ */ new Map();
function getFormatter(locale, decimals, notation = "standard") {
  const key = `${locale}:${decimals}:${notation}`;
  let fmt = _fmtCache.get(key);
  if (!fmt) {
    fmt = new Intl.NumberFormat(locale, {
      minimumFractionDigits: notation === "compact" ? 0 : decimals,
      maximumFractionDigits: decimals,
      ...notation === "compact" && {
        notation: "compact",
        compactDisplay: "short"
      }
    });
    _fmtCache.set(key, fmt);
  }
  return fmt;
}
function formatDirham(amount, options = {}) {
  if (!Number.isFinite(amount)) {
    throw new RangeError(
      `formatDirham: amount must be a finite number, got ${amount}`
    );
  }
  const {
    locale = "en-AE",
    decimals = 2,
    useCode = false,
    separator = "\xA0",
    // non-breaking space
    notation = "standard"
  } = options;
  const symbol = useCode ? DIRHAM_CURRENCY_CODE : DIRHAM_UNICODE;
  let symbolFirst;
  if (options.symbolFirst !== void 0) {
    symbolFirst = options.symbolFirst;
  } else {
    symbolFirst = !locale.startsWith("ar");
  }
  const formatted = getFormatter(locale, decimals, notation).format(amount);
  return symbolFirst ? `${symbol}${separator}${formatted}` : `${formatted}${separator}${symbol}`;
}

// src/web-component/DirhamPriceElement.ts
var SVG_PATH2 = "m88.3 1c0.4 0.6 2.6 3.3 4.7 5.9 15.3 18.2 26.8 47.8 33 85.1 4.1 24.5 4.3 32.2 4.3 125.6v87h-41.8c-38.2 0-42.6-0.2-50.1-1.7-11.8-2.5-24-9.2-32.2-17.8-6.5-6.9-6.3-7.3-5.9 13.6 0.5 17.3 0.7 19.2 3.2 28.6 4 14.9 9.5 26 17.8 35.9 11.3 13.6 22.8 21.2 39.2 26.3 3.5 1 10.9 1.4 37.1 1.6l32.7 0.5v43.3 43.4l-46.1-0.3-46.3-0.3-8-3.2c-9.5-3.8-13.8-6.6-23.1-14.9l-6.8-6.1 0.4 19.1c0.5 17.7 0.6 19.7 3.1 28.7 8.7 31.8 29.7 54.5 57.4 61.9 6.9 1.9 9.6 2 38.5 2.4l30.9 0.4v89.6c0 54.1-0.3 94-0.8 100.8-0.5 6.2-2.1 17.8-3.5 25.9-6.5 37.3-18.2 65.4-35 83.6l-3.4 3.7h169.1c101.1 0 176.7-0.4 187.8-0.9 19.5-1 63-5.3 72.8-7.4 3.1-0.6 8.9-1.5 12.7-2.1 8.1-1.2 21.5-4 40.8-8.9 27.2-6.8 52-15.3 76.3-26.1 7.6-3.4 29.4-14.5 35.2-18 3.1-1.8 6.8-4 8.2-4.7 3.9-2.1 10.4-6.3 19.9-13.1 4.7-3.4 9.4-6.7 10.4-7.4 4.2-2.8 18.7-14.9 25.3-21 25.1-23.1 46.1-48.8 62.4-76.3 2.3-4 5.3-9 6.6-11.1 3.3-5.6 16.9-33.6 18.2-37.8 0.6-1.9 1.4-3.9 1.8-4.3 2.6-3.4 17.6-50.6 19.4-60.9 0.6-3.3 0.9-3.8 3.4-4.3 1.6-0.3 24.9-0.3 51.8-0.1 53.8 0.4 53.8 0.4 65.7 5.9 6.7 3.1 8.7 4.5 16.1 11.2 9.7 8.7 8.8 10.1 8.2-11.7-0.4-12.8-0.9-20.7-1.8-23.9-3.4-12.3-4.2-14.9-7.2-21.1-9.8-21.4-26.2-36.7-47.2-44l-8.2-3-33.4-0.4-33.3-0.5 0.4-11.7c0.4-15.4 0.4-45.9-0.1-61.6l-0.4-12.6 44.6-0.2c38.2-0.2 45.3 0 49.5 1.1 12.6 3.5 21.1 8.3 31.5 17.8l5.8 5.4v-14.8c0-17.6-0.9-25.4-4.5-37-7.1-23.5-21.1-41-41.1-51.8-13-7-13.8-7.2-58.5-7.5-26.2-0.2-39.9-0.6-40.6-1.2-0.6-0.6-1.1-1.6-1.1-2.4 0-0.8-1.5-7.1-3.5-13.9-23.4-82.7-67.1-148.4-131-197.1-8.7-6.7-30-20.8-38.6-25.6-3.3-1.9-6.9-3.9-7.8-4.5-4.2-2.3-28.3-14.1-34.3-16.6-3.6-1.6-8.3-3.6-10.4-4.4-35.3-15.3-94.5-29.8-139.7-34.3-7.4-0.7-17.2-1.8-21.7-2.2-20.4-2.3-48.7-2.6-209.4-2.6-135.8 0-169.9 0.3-169.4 1zm330.7 43.3c33.8 2 54.6 4.6 78.9 10.5 74.2 17.6 126.4 54.8 164.3 117 3.5 5.8 18.3 36 20.5 42.1 10.5 28.3 15.6 45.1 20.1 67.3 1.1 5.4 2.6 12.6 3.3 16 0.7 3.3 1 6.4 0.7 6.7-0.5 0.4-100.9 0.6-223.3 0.5l-222.5-0.2-0.3-128.5c-0.1-70.6 0-129.3 0.3-130.4l0.4-1.9h71.1c39 0 78 0.4 86.5 0.9zm297.5 350.3c0.7 4.3 0.7 77.3 0 80.9l-0.6 2.7-227.5-0.2-227.4-0.3-0.2-42.4c-0.2-23.3 0-42.7 0.2-43.1 0.3-0.5 97.2-0.8 227.7-0.8h227.2zm-10.2 171.7c0.5 1.5-1.9 13.8-6.8 33.8-5.6 22.5-13.2 45.2-20.9 62-3.8 8.6-13.3 27.2-15.6 30.7-1.1 1.6-4.3 6.7-7.1 11.2-18 28.2-43.7 53.9-73 72.9-10.7 6.8-32.7 18.4-38.6 20.2-1.2 0.3-2.5 0.9-3 1.3-0.7 0.6-9.8 4-20.4 7.8-19.5 6.9-56.6 14.4-86.4 17.5-19.3 1.9-22.4 2-96.7 2h-76.9v-129.7-129.8l220.9-0.4c121.5-0.2 221.6-0.5 222.4-0.7 0.9-0.1 1.8 0.5 2.1 1.2z";
var VALID_WEIGHTS2 = /* @__PURE__ */ new Set([
  "thin",
  "extralight",
  "light",
  "regular",
  "medium",
  "semibold",
  "bold",
  "extrabold",
  "black"
]);
var _shadow2, _DirhamPriceElement_instances, render_fn2, escape_fn;
var DirhamPriceElement = class extends HTMLElement {
  constructor() {
    super();
    __privateAdd(this, _DirhamPriceElement_instances);
    __privateAdd(this, _shadow2);
    __privateSet(this, _shadow2, this.attachShadow({ mode: "open" }));
    __privateMethod(this, _DirhamPriceElement_instances, render_fn2).call(this);
  }
  static get observedAttributes() {
    return [
      "amount",
      "locale",
      "decimals",
      "notation",
      "use-code",
      "symbol-size",
      "weight",
      "currency"
    ];
  }
  attributeChangedCallback() {
    __privateMethod(this, _DirhamPriceElement_instances, render_fn2).call(this);
  }
};
_shadow2 = new WeakMap();
_DirhamPriceElement_instances = new WeakSet();
render_fn2 = function() {
  const amountStr = this.getAttribute("amount");
  const amount = amountStr !== null ? Number(amountStr) : 0;
  const locale = this.getAttribute("locale") || "en-AE";
  const decimals = Number(this.getAttribute("decimals") ?? "2");
  const notation = this.getAttribute("notation") === "compact" ? "compact" : "standard";
  const useCode = this.hasAttribute("use-code");
  const symbolSize = this.getAttribute("symbol-size") || "1em";
  const weightAttr = this.getAttribute("weight") || "regular";
  const weight = VALID_WEIGHTS2.has(weightAttr) ? weightAttr : "regular";
  const currency = this.getAttribute("currency") || void 0;
  const symbolFirst = !locale.startsWith("ar");
  let formatted;
  if (!Number.isFinite(amount)) {
    formatted = "\u2014";
  } else {
    formatted = formatDirham(amount, {
      locale,
      decimals,
      useCode: true,
      notation
    }).replace("AED", "").trim();
  }
  if (useCode) {
    const code = currency || "AED";
    const text = symbolFirst ? `${code}\xA0${formatted}` : `${formatted}\xA0${code}`;
    const dirAttr2 = locale.startsWith("ar") ? ' dir="rtl"' : "";
    __privateGet(this, _shadow2).innerHTML = `<span style="white-space:nowrap"${dirAttr2}>${__privateMethod(this, _DirhamPriceElement_instances, escape_fn).call(this, text)}</span>`;
    return;
  }
  const strokeWidth = DIRHAM_STROKE_MAP[weight];
  const strokeAttrs = strokeWidth > 0 ? ` stroke="currentColor" stroke-width="${strokeWidth}" stroke-linejoin="round" paint-order="stroke"` : "";
  const svg = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1000 870" width="${__privateMethod(this, _DirhamPriceElement_instances, escape_fn).call(this, String(symbolSize))}" height="${__privateMethod(this, _DirhamPriceElement_instances, escape_fn).call(this, String(symbolSize))}" fill="currentColor" role="img" aria-label="UAE Dirham" style="display:inline-block;vertical-align:middle"><path d="${SVG_PATH2}"${strokeAttrs}/></svg>`;
  const dirAttr = locale.startsWith("ar") ? ' dir="rtl"' : "";
  const parts = symbolFirst ? `${svg}\xA0${__privateMethod(this, _DirhamPriceElement_instances, escape_fn).call(this, formatted)}` : `${__privateMethod(this, _DirhamPriceElement_instances, escape_fn).call(this, formatted)}\xA0${svg}`;
  __privateGet(this, _shadow2).innerHTML = `<span style="white-space:nowrap"${dirAttr}>${parts}</span>`;
};
escape_fn = function(str) {
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
};
if (typeof customElements !== "undefined" && !customElements.get("dirham-price")) {
  customElements.define("dirham-price", DirhamPriceElement);
}

// src/web-component/DirhamInputElement.ts
var SVG_PATH3 = "m88.3 1c0.4 0.6 2.6 3.3 4.7 5.9 15.3 18.2 26.8 47.8 33 85.1 4.1 24.5 4.3 32.2 4.3 125.6v87h-41.8c-38.2 0-42.6-0.2-50.1-1.7-11.8-2.5-24-9.2-32.2-17.8-6.5-6.9-6.3-7.3-5.9 13.6 0.5 17.3 0.7 19.2 3.2 28.6 4 14.9 9.5 26 17.8 35.9 11.3 13.6 22.8 21.2 39.2 26.3 3.5 1 10.9 1.4 37.1 1.6l32.7 0.5v43.3 43.4l-46.1-0.3-46.3-0.3-8-3.2c-9.5-3.8-13.8-6.6-23.1-14.9l-6.8-6.1 0.4 19.1c0.5 17.7 0.6 19.7 3.1 28.7 8.7 31.8 29.7 54.5 57.4 61.9 6.9 1.9 9.6 2 38.5 2.4l30.9 0.4v89.6c0 54.1-0.3 94-0.8 100.8-0.5 6.2-2.1 17.8-3.5 25.9-6.5 37.3-18.2 65.4-35 83.6l-3.4 3.7h169.1c101.1 0 176.7-0.4 187.8-0.9 19.5-1 63-5.3 72.8-7.4 3.1-0.6 8.9-1.5 12.7-2.1 8.1-1.2 21.5-4 40.8-8.9 27.2-6.8 52-15.3 76.3-26.1 7.6-3.4 29.4-14.5 35.2-18 3.1-1.8 6.8-4 8.2-4.7 3.9-2.1 10.4-6.3 19.9-13.1 4.7-3.4 9.4-6.7 10.4-7.4 4.2-2.8 18.7-14.9 25.3-21 25.1-23.1 46.1-48.8 62.4-76.3 2.3-4 5.3-9 6.6-11.1 3.3-5.6 16.9-33.6 18.2-37.8 0.6-1.9 1.4-3.9 1.8-4.3 2.6-3.4 17.6-50.6 19.4-60.9 0.6-3.3 0.9-3.8 3.4-4.3 1.6-0.3 24.9-0.3 51.8-0.1 53.8 0.4 53.8 0.4 65.7 5.9 6.7 3.1 8.7 4.5 16.1 11.2 9.7 8.7 8.8 10.1 8.2-11.7-0.4-12.8-0.9-20.7-1.8-23.9-3.4-12.3-4.2-14.9-7.2-21.1-9.8-21.4-26.2-36.7-47.2-44l-8.2-3-33.4-0.4-33.3-0.5 0.4-11.7c0.4-15.4 0.4-45.9-0.1-61.6l-0.4-12.6 44.6-0.2c38.2-0.2 45.3 0 49.5 1.1 12.6 3.5 21.1 8.3 31.5 17.8l5.8 5.4v-14.8c0-17.6-0.9-25.4-4.5-37-7.1-23.5-21.1-41-41.1-51.8-13-7-13.8-7.2-58.5-7.5-26.2-0.2-39.9-0.6-40.6-1.2-0.6-0.6-1.1-1.6-1.1-2.4 0-0.8-1.5-7.1-3.5-13.9-23.4-82.7-67.1-148.4-131-197.1-8.7-6.7-30-20.8-38.6-25.6-3.3-1.9-6.9-3.9-7.8-4.5-4.2-2.3-28.3-14.1-34.3-16.6-3.6-1.6-8.3-3.6-10.4-4.4-35.3-15.3-94.5-29.8-139.7-34.3-7.4-0.7-17.2-1.8-21.7-2.2-20.4-2.3-48.7-2.6-209.4-2.6-135.8 0-169.9 0.3-169.4 1zm330.7 43.3c33.8 2 54.6 4.6 78.9 10.5 74.2 17.6 126.4 54.8 164.3 117 3.5 5.8 18.3 36 20.5 42.1 10.5 28.3 15.6 45.1 20.1 67.3 1.1 5.4 2.6 12.6 3.3 16 0.7 3.3 1 6.4 0.7 6.7-0.5 0.4-100.9 0.6-223.3 0.5l-222.5-0.2-0.3-128.5c-0.1-70.6 0-129.3 0.3-130.4l0.4-1.9h71.1c39 0 78 0.4 86.5 0.9zm297.5 350.3c0.7 4.3 0.7 77.3 0 80.9l-0.6 2.7-227.5-0.2-227.4-0.3-0.2-42.4c-0.2-23.3 0-42.7 0.2-43.1 0.3-0.5 97.2-0.8 227.7-0.8h227.2zm-10.2 171.7c0.5 1.5-1.9 13.8-6.8 33.8-5.6 22.5-13.2 45.2-20.9 62-3.8 8.6-13.3 27.2-15.6 30.7-1.1 1.6-4.3 6.7-7.1 11.2-18 28.2-43.7 53.9-73 72.9-10.7 6.8-32.7 18.4-38.6 20.2-1.2 0.3-2.5 0.9-3 1.3-0.7 0.6-9.8 4-20.4 7.8-19.5 6.9-56.6 14.4-86.4 17.5-19.3 1.9-22.4 2-96.7 2h-76.9v-129.7-129.8l220.9-0.4c121.5-0.2 221.6-0.5 222.4-0.7 0.9-0.1 1.8 0.5 2.1 1.2z";
var VALID_WEIGHTS3 = /* @__PURE__ */ new Set([
  "thin",
  "extralight",
  "light",
  "regular",
  "medium",
  "semibold",
  "bold",
  "extrabold",
  "black"
]);
var _fmtCache2 = /* @__PURE__ */ new Map();
function getInputFormatter(locale, decimals) {
  const key = `wc-input:${locale}:${decimals}`;
  let fmt = _fmtCache2.get(key);
  if (!fmt) {
    fmt = new Intl.NumberFormat(locale, {
      minimumFractionDigits: 0,
      maximumFractionDigits: decimals
    });
    _fmtCache2.set(key, fmt);
  }
  return fmt;
}
function stripFormatting(raw) {
  let cleaned = raw.replace(
    /[\u0660-\u0669]/g,
    (d) => String(d.charCodeAt(0) - 1632)
  );
  cleaned = cleaned.replace(/\u066B/g, ".");
  cleaned = cleaned.replace(/[,\s\u00A0\u066C]/g, "");
  cleaned = cleaned.replace(/[\u20C3]/g, "");
  cleaned = cleaned.replace(/AED/gi, "");
  cleaned = cleaned.replace(/د\.إ/g, "");
  return cleaned.trim();
}
function escapeHTML(str) {
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}
var _shadow3, _input, _numericValue, _isFocused, _DirhamInputElement_instances, getConfig_fn, clamp_fn, formatDisplay_fn, render_fn3, attachListeners_fn;
var DirhamInputElement = class extends HTMLElement {
  constructor() {
    super();
    __privateAdd(this, _DirhamInputElement_instances);
    __privateAdd(this, _shadow3);
    __privateAdd(this, _input, null);
    __privateAdd(this, _numericValue, null);
    __privateAdd(this, _isFocused, false);
    __privateSet(this, _shadow3, this.attachShadow({ mode: "open" }));
    __privateMethod(this, _DirhamInputElement_instances, render_fn3).call(this);
    __privateMethod(this, _DirhamInputElement_instances, attachListeners_fn).call(this);
  }
  static get observedAttributes() {
    return [
      "value",
      "locale",
      "decimals",
      "min",
      "max",
      "placeholder",
      "disabled",
      "readonly",
      "show-symbol",
      "use-code",
      "symbol-size",
      "weight"
    ];
  }
  get numericValue() {
    return __privateGet(this, _numericValue);
  }
  attributeChangedCallback() {
    if (!__privateGet(this, _isFocused)) {
      __privateMethod(this, _DirhamInputElement_instances, render_fn3).call(this);
      __privateMethod(this, _DirhamInputElement_instances, attachListeners_fn).call(this);
    }
  }
};
_shadow3 = new WeakMap();
_input = new WeakMap();
_numericValue = new WeakMap();
_isFocused = new WeakMap();
_DirhamInputElement_instances = new WeakSet();
getConfig_fn = function() {
  const locale = this.getAttribute("locale") || "en-AE";
  const decimals = Number(this.getAttribute("decimals") ?? "2");
  const min = this.hasAttribute("min") ? Number(this.getAttribute("min")) : void 0;
  const max = this.hasAttribute("max") ? Number(this.getAttribute("max")) : void 0;
  const showSymbol = !this.hasAttribute("show-symbol") || this.getAttribute("show-symbol") !== "false";
  const useCode = this.hasAttribute("use-code");
  const symbolSize = this.getAttribute("symbol-size") || "1em";
  const weightAttr = this.getAttribute("weight") || "regular";
  const weight = VALID_WEIGHTS3.has(weightAttr) ? weightAttr : "regular";
  const disabled = this.hasAttribute("disabled");
  const readonly = this.hasAttribute("readonly");
  const isRTL = locale.startsWith("ar");
  return {
    locale,
    decimals,
    min,
    max,
    showSymbol,
    useCode,
    symbolSize,
    weight,
    disabled,
    readonly,
    isRTL
  };
};
clamp_fn = function(v) {
  const { min, max } = __privateMethod(this, _DirhamInputElement_instances, getConfig_fn).call(this);
  let result = v;
  if (min !== void 0 && result < min) result = min;
  if (max !== void 0 && result > max) result = max;
  return result;
};
formatDisplay_fn = function(value) {
  if (value === null || !Number.isFinite(value)) return "";
  const { locale, decimals } = __privateMethod(this, _DirhamInputElement_instances, getConfig_fn).call(this);
  return getInputFormatter(locale, decimals).format(value);
};
render_fn3 = function() {
  const config = __privateMethod(this, _DirhamInputElement_instances, getConfig_fn).call(this);
  const valAttr = this.getAttribute("value");
  if (valAttr !== null && !__privateGet(this, _isFocused)) {
    const parsed = Number(valAttr);
    __privateSet(this, _numericValue, Number.isFinite(parsed) ? parsed : null);
  }
  const displayVal = __privateGet(this, _isFocused) ? void 0 : __privateMethod(this, _DirhamInputElement_instances, formatDisplay_fn).call(this, __privateGet(this, _numericValue));
  const strokeWidth = DIRHAM_STROKE_MAP[config.weight];
  const strokeAttrs = strokeWidth > 0 ? ` stroke="currentColor" stroke-width="${strokeWidth}" stroke-linejoin="round" paint-order="stroke"` : "";
  const symbolHTML = config.showSymbol ? config.useCode ? `<span style="user-select:none;flex-shrink:0;${config.isRTL ? "margin-left" : "margin-right"}:0.5em" aria-hidden="true">AED</span>` : `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1000 870" width="${escapeHTML(String(config.symbolSize))}" height="${escapeHTML(String(config.symbolSize))}" fill="currentColor" role="img" aria-hidden="true" style="display:inline-block;vertical-align:middle;flex-shrink:0;${config.isRTL ? "margin-left" : "margin-right"}:0.5em"><path d="${SVG_PATH3}"${strokeAttrs}/></svg>` : "";
  const placeholder = this.getAttribute("placeholder") || getInputFormatter(config.locale, config.decimals).format(0);
  const html = `
<style>
:host { display: inline-flex; align-items: center; direction: ${config.isRTL ? "rtl" : "ltr"}; }
input { min-width: 0; flex: 1; border: none; background: transparent; font: inherit; color: inherit; outline: none; padding: 0; direction: ${config.isRTL ? "rtl" : "ltr"}; }
</style>
${config.isRTL ? "" : symbolHTML}
<input type="text" inputmode="decimal" aria-label="Amount in AED" placeholder="${escapeHTML(placeholder)}"${config.disabled ? " disabled" : ""}${config.readonly ? " readonly" : ""}${displayVal !== void 0 ? ` value="${escapeHTML(displayVal)}"` : ""}/>
${config.isRTL ? symbolHTML : ""}`;
  if (!__privateGet(this, _isFocused)) {
    __privateGet(this, _shadow3).innerHTML = html;
    __privateSet(this, _input, __privateGet(this, _shadow3).querySelector("input"));
  }
};
attachListeners_fn = function() {
  const input = __privateGet(this, _shadow3).querySelector("input");
  if (!input || input === __privateGet(this, _input)) return;
  __privateSet(this, _input, input);
  input.addEventListener("focus", () => {
    __privateSet(this, _isFocused, true);
    const plain = __privateGet(this, _numericValue) !== null && Number.isFinite(__privateGet(this, _numericValue)) ? __privateGet(this, _numericValue).toString() : "";
    input.value = plain;
    setTimeout(() => input.select(), 0);
  });
  input.addEventListener("blur", () => {
    __privateSet(this, _isFocused, false);
    const stripped = stripFormatting(input.value);
    const { decimals } = __privateMethod(this, _DirhamInputElement_instances, getConfig_fn).call(this);
    if (stripped === "" || stripped === "." || stripped === "-") {
      __privateSet(this, _numericValue, null);
    } else {
      let parsed = Number.parseFloat(stripped);
      if (Number.isNaN(parsed)) {
        __privateSet(this, _numericValue, null);
      } else {
        const factor = 10 ** decimals;
        parsed = Math.round(parsed * factor) / factor;
        parsed = __privateMethod(this, _DirhamInputElement_instances, clamp_fn).call(this, parsed);
        __privateSet(this, _numericValue, parsed);
      }
    }
    __privateMethod(this, _DirhamInputElement_instances, render_fn3).call(this);
    __privateMethod(this, _DirhamInputElement_instances, attachListeners_fn).call(this);
    this.dispatchEvent(
      new CustomEvent("dirham-change", {
        detail: { value: __privateGet(this, _numericValue) },
        bubbles: true,
        composed: true
      })
    );
  });
  input.addEventListener("input", () => {
    const cleaned = stripFormatting(input.value);
    if (cleaned !== "" && !/^-?\d*\.?\d*$/.test(cleaned)) {
      const prev = __privateGet(this, _numericValue) !== null ? __privateGet(this, _numericValue).toString() : "";
      input.value = prev;
      return;
    }
    const { decimals } = __privateMethod(this, _DirhamInputElement_instances, getConfig_fn).call(this);
    const dotIndex = cleaned.indexOf(".");
    if (dotIndex !== -1 && cleaned.slice(dotIndex + 1).length > decimals) {
      const prev = __privateGet(this, _numericValue) !== null ? __privateGet(this, _numericValue).toString() : "";
      input.value = prev;
      return;
    }
    input.value = cleaned;
  });
  input.addEventListener("paste", (e) => {
    e.preventDefault();
    const pasted = e.clipboardData?.getData("text/plain") ?? "";
    const stripped = stripFormatting(pasted);
    if (stripped === "") return;
    const { decimals } = __privateMethod(this, _DirhamInputElement_instances, getConfig_fn).call(this);
    let final = stripped;
    const dotIndex = stripped.indexOf(".");
    if (dotIndex !== -1) {
      final = `${stripped.slice(0, dotIndex)}.${stripped.slice(dotIndex + 1, dotIndex + 1 + decimals)}`;
    }
    if (!/^-?\d*\.?\d*$/.test(final)) return;
    input.value = final;
    const parsed = Number.parseFloat(final);
    if (!Number.isNaN(parsed)) {
      const factor = 10 ** decimals;
      __privateSet(this, _numericValue, __privateMethod(this, _DirhamInputElement_instances, clamp_fn).call(this, Math.round(parsed * factor) / factor));
    }
  });
};
if (typeof customElements !== "undefined" && !customElements.get("dirham-input")) {
  customElements.define("dirham-input", DirhamInputElement);
}

// src/web-component/AnimatedDirhamPriceElement.ts
var SVG_PATH4 = "m88.3 1c0.4 0.6 2.6 3.3 4.7 5.9 15.3 18.2 26.8 47.8 33 85.1 4.1 24.5 4.3 32.2 4.3 125.6v87h-41.8c-38.2 0-42.6-0.2-50.1-1.7-11.8-2.5-24-9.2-32.2-17.8-6.5-6.9-6.3-7.3-5.9 13.6 0.5 17.3 0.7 19.2 3.2 28.6 4 14.9 9.5 26 17.8 35.9 11.3 13.6 22.8 21.2 39.2 26.3 3.5 1 10.9 1.4 37.1 1.6l32.7 0.5v43.3 43.4l-46.1-0.3-46.3-0.3-8-3.2c-9.5-3.8-13.8-6.6-23.1-14.9l-6.8-6.1 0.4 19.1c0.5 17.7 0.6 19.7 3.1 28.7 8.7 31.8 29.7 54.5 57.4 61.9 6.9 1.9 9.6 2 38.5 2.4l30.9 0.4v89.6c0 54.1-0.3 94-0.8 100.8-0.5 6.2-2.1 17.8-3.5 25.9-6.5 37.3-18.2 65.4-35 83.6l-3.4 3.7h169.1c101.1 0 176.7-0.4 187.8-0.9 19.5-1 63-5.3 72.8-7.4 3.1-0.6 8.9-1.5 12.7-2.1 8.1-1.2 21.5-4 40.8-8.9 27.2-6.8 52-15.3 76.3-26.1 7.6-3.4 29.4-14.5 35.2-18 3.1-1.8 6.8-4 8.2-4.7 3.9-2.1 10.4-6.3 19.9-13.1 4.7-3.4 9.4-6.7 10.4-7.4 4.2-2.8 18.7-14.9 25.3-21 25.1-23.1 46.1-48.8 62.4-76.3 2.3-4 5.3-9 6.6-11.1 3.3-5.6 16.9-33.6 18.2-37.8 0.6-1.9 1.4-3.9 1.8-4.3 2.6-3.4 17.6-50.6 19.4-60.9 0.6-3.3 0.9-3.8 3.4-4.3 1.6-0.3 24.9-0.3 51.8-0.1 53.8 0.4 53.8 0.4 65.7 5.9 6.7 3.1 8.7 4.5 16.1 11.2 9.7 8.7 8.8 10.1 8.2-11.7-0.4-12.8-0.9-20.7-1.8-23.9-3.4-12.3-4.2-14.9-7.2-21.1-9.8-21.4-26.2-36.7-47.2-44l-8.2-3-33.4-0.4-33.3-0.5 0.4-11.7c0.4-15.4 0.4-45.9-0.1-61.6l-0.4-12.6 44.6-0.2c38.2-0.2 45.3 0 49.5 1.1 12.6 3.5 21.1 8.3 31.5 17.8l5.8 5.4v-14.8c0-17.6-0.9-25.4-4.5-37-7.1-23.5-21.1-41-41.1-51.8-13-7-13.8-7.2-58.5-7.5-26.2-0.2-39.9-0.6-40.6-1.2-0.6-0.6-1.1-1.6-1.1-2.4 0-0.8-1.5-7.1-3.5-13.9-23.4-82.7-67.1-148.4-131-197.1-8.7-6.7-30-20.8-38.6-25.6-3.3-1.9-6.9-3.9-7.8-4.5-4.2-2.3-28.3-14.1-34.3-16.6-3.6-1.6-8.3-3.6-10.4-4.4-35.3-15.3-94.5-29.8-139.7-34.3-7.4-0.7-17.2-1.8-21.7-2.2-20.4-2.3-48.7-2.6-209.4-2.6-135.8 0-169.9 0.3-169.4 1zm330.7 43.3c33.8 2 54.6 4.6 78.9 10.5 74.2 17.6 126.4 54.8 164.3 117 3.5 5.8 18.3 36 20.5 42.1 10.5 28.3 15.6 45.1 20.1 67.3 1.1 5.4 2.6 12.6 3.3 16 0.7 3.3 1 6.4 0.7 6.7-0.5 0.4-100.9 0.6-223.3 0.5l-222.5-0.2-0.3-128.5c-0.1-70.6 0-129.3 0.3-130.4l0.4-1.9h71.1c39 0 78 0.4 86.5 0.9zm297.5 350.3c0.7 4.3 0.7 77.3 0 80.9l-0.6 2.7-227.5-0.2-227.4-0.3-0.2-42.4c-0.2-23.3 0-42.7 0.2-43.1 0.3-0.5 97.2-0.8 227.7-0.8h227.2zm-10.2 171.7c0.5 1.5-1.9 13.8-6.8 33.8-5.6 22.5-13.2 45.2-20.9 62-3.8 8.6-13.3 27.2-15.6 30.7-1.1 1.6-4.3 6.7-7.1 11.2-18 28.2-43.7 53.9-73 72.9-10.7 6.8-32.7 18.4-38.6 20.2-1.2 0.3-2.5 0.9-3 1.3-0.7 0.6-9.8 4-20.4 7.8-19.5 6.9-56.6 14.4-86.4 17.5-19.3 1.9-22.4 2-96.7 2h-76.9v-129.7-129.8l220.9-0.4c121.5-0.2 221.6-0.5 222.4-0.7 0.9-0.1 1.8 0.5 2.1 1.2z";
var VALID_WEIGHTS4 = /* @__PURE__ */ new Set([
  "thin",
  "extralight",
  "light",
  "regular",
  "medium",
  "semibold",
  "bold",
  "extrabold",
  "black"
]);
var EASING_FNS = {
  linear: (t) => t,
  easeIn: (t) => t * t,
  easeOut: (t) => t * (2 - t),
  easeInOut: (t) => t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t
};
var _shadow4, _currentAmount, _targetAmount, _frameId, _AnimatedDirhamPriceElement_instances, animateTo_fn, escape_fn2, renderFrame_fn;
var AnimatedDirhamPriceElement = class extends HTMLElement {
  constructor() {
    super();
    __privateAdd(this, _AnimatedDirhamPriceElement_instances);
    __privateAdd(this, _shadow4);
    __privateAdd(this, _currentAmount, 0);
    __privateAdd(this, _targetAmount, 0);
    __privateAdd(this, _frameId, 0);
    __privateSet(this, _shadow4, this.attachShadow({ mode: "open" }));
    const amountStr = this.getAttribute("amount");
    __privateSet(this, _currentAmount, amountStr !== null ? Number(amountStr) : 0);
    __privateSet(this, _targetAmount, __privateGet(this, _currentAmount));
    __privateMethod(this, _AnimatedDirhamPriceElement_instances, renderFrame_fn).call(this, __privateGet(this, _currentAmount));
  }
  static get observedAttributes() {
    return [
      "amount",
      "duration",
      "easing",
      "locale",
      "decimals",
      "notation",
      "use-code",
      "symbol-size",
      "weight"
    ];
  }
  attributeChangedCallback(name) {
    if (name === "amount") {
      const newTarget = Number(this.getAttribute("amount") ?? "0");
      if (Number.isFinite(newTarget) && newTarget !== __privateGet(this, _targetAmount)) {
        __privateMethod(this, _AnimatedDirhamPriceElement_instances, animateTo_fn).call(this, newTarget);
      }
    } else {
      __privateMethod(this, _AnimatedDirhamPriceElement_instances, renderFrame_fn).call(this, __privateGet(this, _currentAmount));
    }
  }
};
_shadow4 = new WeakMap();
_currentAmount = new WeakMap();
_targetAmount = new WeakMap();
_frameId = new WeakMap();
_AnimatedDirhamPriceElement_instances = new WeakSet();
animateTo_fn = function(target) {
  if (__privateGet(this, _frameId)) cancelAnimationFrame(__privateGet(this, _frameId));
  const from = __privateGet(this, _currentAmount);
  __privateSet(this, _targetAmount, target);
  const duration = Number(this.getAttribute("duration") ?? "600");
  const easingName = this.getAttribute("easing") || "easeOut";
  const easeFn = EASING_FNS[easingName] ?? EASING_FNS.easeOut;
  const start = performance.now();
  const animate = (now) => {
    const elapsed = now - start;
    const progress = Math.min(elapsed / duration, 1);
    const eased = easeFn(progress);
    __privateSet(this, _currentAmount, from + (target - from) * eased);
    __privateMethod(this, _AnimatedDirhamPriceElement_instances, renderFrame_fn).call(this, __privateGet(this, _currentAmount));
    if (progress < 1) {
      __privateSet(this, _frameId, requestAnimationFrame(animate));
    } else {
      __privateSet(this, _currentAmount, target);
      __privateMethod(this, _AnimatedDirhamPriceElement_instances, renderFrame_fn).call(this, target);
      __privateSet(this, _frameId, 0);
    }
  };
  __privateSet(this, _frameId, requestAnimationFrame(animate));
};
escape_fn2 = function(str) {
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
};
renderFrame_fn = function(amount) {
  const locale = this.getAttribute("locale") || "en-AE";
  const decimals = Number(this.getAttribute("decimals") ?? "2");
  const notation = this.getAttribute("notation") === "compact" ? "compact" : "standard";
  const useCode = this.hasAttribute("use-code");
  const symbolSize = this.getAttribute("symbol-size") || "1em";
  const weightAttr = this.getAttribute("weight") || "regular";
  const weight = VALID_WEIGHTS4.has(weightAttr) ? weightAttr : "regular";
  const symbolFirst = !locale.startsWith("ar");
  const isRTL = locale.startsWith("ar");
  const dirAttr = isRTL ? ' dir="rtl"' : "";
  let formatted;
  if (!Number.isFinite(amount)) {
    formatted = "\u2014";
  } else {
    formatted = formatDirham(amount, {
      locale,
      decimals,
      useCode: true,
      notation
    }).replace("AED", "").trim();
  }
  if (useCode) {
    const text = symbolFirst ? `AED\xA0${formatted}` : `${formatted}\xA0AED`;
    __privateGet(this, _shadow4).innerHTML = `<span style="white-space:nowrap"${dirAttr} aria-live="polite" aria-atomic="true">${__privateMethod(this, _AnimatedDirhamPriceElement_instances, escape_fn2).call(this, text)}</span>`;
    return;
  }
  const strokeWidth = DIRHAM_STROKE_MAP[weight];
  const strokeAttrs = strokeWidth > 0 ? ` stroke="currentColor" stroke-width="${strokeWidth}" stroke-linejoin="round" paint-order="stroke"` : "";
  const svg = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1000 870" width="${__privateMethod(this, _AnimatedDirhamPriceElement_instances, escape_fn2).call(this, String(symbolSize))}" height="${__privateMethod(this, _AnimatedDirhamPriceElement_instances, escape_fn2).call(this, String(symbolSize))}" fill="currentColor" role="img" aria-label="UAE Dirham" style="display:inline-block;vertical-align:middle"><path d="${SVG_PATH4}"${strokeAttrs}/></svg>`;
  const parts = symbolFirst ? `${svg}\xA0${__privateMethod(this, _AnimatedDirhamPriceElement_instances, escape_fn2).call(this, formatted)}` : `${__privateMethod(this, _AnimatedDirhamPriceElement_instances, escape_fn2).call(this, formatted)}\xA0${svg}`;
  __privateGet(this, _shadow4).innerHTML = `<span style="white-space:nowrap"${dirAttr} aria-live="polite" aria-atomic="true">${parts}</span>`;
};
if (typeof customElements !== "undefined" && !customElements.get("dirham-animated-price")) {
  customElements.define("dirham-animated-price", AnimatedDirhamPriceElement);
}

exports.AnimatedDirhamPriceElement = AnimatedDirhamPriceElement;
exports.DirhamInputElement = DirhamInputElement;
exports.DirhamPriceElement = DirhamPriceElement;
exports.DirhamSymbolElement = DirhamSymbolElement;
